### PR TITLE
FEAT: Siwave xtalk scan config file support

### PR DESCRIPTION
**Pull request by svandenb-dev**
Adding siwave xml config file creation support

### CODE REVIEW

File: src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    ):
+        if name and name not in self.nets:
+            net = SingleEndedNet()
+            net.name = name
+            net.next_warning_threshold = next_warning_threshold
+            net.next_violation_threshold = next_violation_threshold
+            net.fext_warning_threshold = fext_warning_threshold_warning
+            net.fext_violation_threshold = fext_violation_threshold
+            self.nets[name] = net
+        else:
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py:57]: Can you add a log error if name is already in self.nets ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/impedance_scan_config.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    def add_single_ended_net(self, name, nominal_impedance=50.0, warning_threshold=17.0, violation_threshold=32.0):
+        if name and name not in self.nets:
+            net = SingleEndedNet()
+            net.name = name
+            net.nominal_impedance = nominal_impedance
+            net.warning_threshold = warning_threshold
+            net.violation_threshold = violation_threshold
+            self.nets[name] = net
+        else:
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/impedance_scan_config.py:49]: Same as previous comment (log)

File: src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+from pyedb.misc.siw_feature_config.xtalk_scan.net import SingleEndedNet
+
+
+class CrosstalkFrequency:
+    def __init__(self):
+        self.min_transmission_line_segment_length = "0.25mm"
+        self.frequency = "2e9Hz"
+        self.nets = {}
+
+    def write_wml(self, parent):
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py:33]: Don't you want to call this method `write_xml` ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/impedance_scan_config.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+from pyedb.misc.siw_feature_config.xtalk_scan.net import SingleEndedNet
+
+
+class ImpedanceScan:
+    def __init__(self):
+        self.min_transmission_line_segment_length = "0.25mm"
+        self.frequency = "2e9Hz"
+        self.nets = {}
+
+    def write_wml(self, parent):
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/impedance_scan_config.py:33]: Method name ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/net.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        if self.next_violation_threshold is not None:
+            net.set("NEXTViolationThreshold", str(self.next_violation_threshold))
+        if self.driver_rise_time is not None:
+            net.set("DriverRiseTime", str(self.driver_rise_time))
+        if self.voltage is not None:
+            net.set("Voltage", str(self.voltage))
+        if self.driver_impedance is not None:
+            net.set("DriverImpedance", str(self.driver_impedance))
+        if self.termination_impedance is not None:
+            net.set("TerminationImpedance", str(self.termination_impedance))
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/net.py:66]: You are not writting the xml file

File: src/pyedb/misc/siw_feature_config/xtalk_scan/pins.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        self.voltage = None
+        self.driver_impedance = None
+
+    def write_xml(self, parent):
+        pin = ET.SubElement(parent, "Pin")
+        pin.set("Name", self.name)
+        pin.set("RefDes", self.ref_des)
+        pin.set("DriverRiseTime", str(self.driver_rise_time))
+        pin.set("Voltage", str(self.voltage))
+        pin.set("DriverImpedance", str(self.driver_impedance))
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/pins.py:41]: no xml file writting ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/pins.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    def __init__(self):
+        self.name = None
+        self.ref_des = None
+        self.receiver_impedance = None
+
+    def write_xml(self, parent):
+        pin = ET.SubElement(parent, "Pin")
+        pin.set("Name", self.name)
+        pin.set("Name", self.name)
+        pin.set("ReceiverImpedance", str(self.receiver_impedance))
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/pins.py:54]: no xml file writting ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/scan_config.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            self.scan_type = ScanType.IMPEDANCE
+        elif scan_type == "frequency_xtalk":
+            self.scan_type = ScanType.FREQ_XTALK
+        elif scan_type == "time_xtalk":
+            self.scan_type = ScanType.TIME_XTALK
+        self.impedance_scan = ImpedanceScan()
+        self.frequency_xtalk_scan = CrosstalkFrequency()
+        self.time_xtalk_scan = CrossTalkTime()
+
+    def write_wml(self):
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/scan_config.py:54]: Name issue ?

File: src/pyedb/misc/siw_feature_config/xtalk_scan/td_xtalk_config.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        self.driver_pins.append(pin)
+
+    def add_receiver_pin(self, name, ref_des, impedance):
+        pin = ReceiverPin()
+        pin.name = name
+        pin.ref_des = ref_des
+        pin.receiver_impedance = impedance
+        self.receiver_pins.append(pin)
+
+    def write_wml(self, parent):
```
Comment (SMoraisAnsys) [inline src/pyedb/misc/siw_feature_config/xtalk_scan/td_xtalk_config.py:69]: name

**Review by SMoraisAnsys — CHANGES_REQUESTED** [review]
I left a lot of comments and some might be missing.
If I didn't understood things correctly, let me know !

File: src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py (hunk 3 above)
Comment (svandenb-dev) [reply src/pyedb/misc/siw_feature_config/xtalk_scan/fd_xtalk_scan_config.py:33]: oups yes

File: tests/legacy/system/test_edb_crosstalk_scanner.py
Comment (ring630) [inline tests/legacy/system/test_edb_crosstalk_scanner.py:1]: @svandenb-dev can you rename the file as test_siwave_features.py?

File: tests/legacy/system/test_edb_crosstalk_scanner.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+import pytest
+
+pytestmark = [pytest.mark.system, pytest.mark.legacy]
+
+
+class TestClass:
+    @pytest.fixture(autouse=True)
+    def init(self, legacy_edb_app, local_scratch, target_path, target_path2, target_path4):
+        self.edbapp = legacy_edb_app
```
Comment (ring630) [inline tests/legacy/system/test_edb_crosstalk_scanner.py:27]: Can you use edb_examples.get_si_verse()

File: src/pyedb/dotnet/edb_core/siwave.py [hunk 11]
```diff
@@ -1488,6 +1489,19 @@ def icepak_use_minimal_comp_defaults(self):
         _, value = cell.GetProductProperty(siwave_id, 422, "")
         return bool(value)
 
+    def create_crosstalk_config_file(self, scan_type="impedance"):
```
Comment (ring630) [inline src/pyedb/dotnet/edb_core/siwave.py:1492]: Can you use a create a another method to create impedance scan config file?

File: tests/legacy/system/test_edb_crosstalk_scanner.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        self.target_path4 = target_path4
+
+    def test_create_impedance_scan(self):
+        xtalk_scan = self.edbapp.siwave.create_crosstalk_config_file(scan_type="impedance")
+        for net in list(self.edbapp.nets.signal.keys()):
+            xtalk_scan.impedance_scan.add_single_ended_net(
+                name=net, nominal_impedance=45.0, warning_threshold=40.0, violation_threshold=30.0
+            )
+        xtalk_scan.file_path = os.path.join(self.local_scratch.path, "test_impedance_scan.xml")
+        assert xtalk_scan.write_xml()
```
Comment (ring630) [inline tests/legacy/system/test_edb_crosstalk_scanner.py:40]: It only checks if the xml file exists. Can you also check if content of the file is valid?

File: tests/legacy/system/test_edb_crosstalk_scanner.py
Comment (svandenb-dev) [reply tests/legacy/system/test_edb_crosstalk_scanner.py:1]: Done

File: src/pyedb/dotnet/edb_core/siwave.py (hunk 11 above)
Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/siwave.py:1492]: it is also supporting impedance scan

Comment (svandenb-dev) [reply src/pyedb/dotnet/edb_core/siwave.py]: changed the name to create_impedance_crosstalk_scan

File: tests/legacy/system/test_edb_crosstalk_scanner.py (hunk 12 above)
Comment (svandenb-dev) [reply tests/legacy/system/test_edb_crosstalk_scanner.py]: done

**Review by SMoraisAnsys — APPROVED** [review]

**Pull request merged**